pkg/server/v3/delta_test.go

```diff
@@ -0,0 +1,413 @@
+			},
+		},
+		rsrc.RuntimeType: {
+			&cache.RawDeltaResponse{
+				SystemVersionInfo: "6",
+				Resources:         []types.Resource{runtime},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.RuntimeType},
+			},
+		},
+		// Pass-through type (xDS does not exist for this type)
+		opaqueType: {
+			&cache.RawDeltaResponse{
+				SystemVersionInfo: "7",
+				Resources:         []types.Resource{opaque},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: opaqueType},
+			},
+		},
+	}
+}
+
+func TestDeltaResponseHandlersWildcard(t *testing.T) {
+	for _, typ := range testTypes {
+		t.Run(typ, func(t *testing.T) {
+			config := makeMockConfigWatcher()
+			config.deltaResponses = makeDeltaResponses()
+			s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+
+			resp := makeMockDeltaStream(t)
+			// This is a wildcard request since we don't specify a list of resource subscriptions
+			resp.recv <- &discovery.DeltaDiscoveryRequest{Node: node, TypeUrl: typ}
+
+			go func() {
+				var err error
+				switch typ {
+				case rsrc.EndpointType:
+					err = s.DeltaEndpoints(resp)
+				case rsrc.ClusterType:
+					err = s.DeltaClusters(resp)
+				case rsrc.RouteType:
+					err = s.DeltaRoutes(resp)
+				case rsrc.ListenerType:
+					err = s.DeltaListeners(resp)
+				case rsrc.SecretType:
+					err = s.DeltaSecrets(resp)
+				case rsrc.RuntimeType:
+					err = s.DeltaRuntime(resp)
+				case opaqueType:
+					err = s.DeltaAggregatedResources(resp)
+				}
+
+				if err != nil {
+					t.Errorf("Delta() => got \"%v\", want no error", err)
+				}
+			}()
+
+			select {
+			case res := <-resp.sent:
+				close(resp.recv)
+
+				// We should only have 7 watch channels initialized since that is the base map length
+				if config.deltaCounts[typ] != 1 {
+					t.Errorf("watch counts for typ: %s => got %v, want 1", typ, config.deltaCounts[typ])
+				}
+
+				if v := res.GetSystemVersionInfo(); v != "" {
+					t.Errorf("expected emtpy version on initial request, got %s", v)
+				}
+			case <-time.After(1 * time.Second):
+				t.Fatalf("got no response")
+			}
+		})
+	}
+}
+
+func TestDeltaResponseHandlers(t *testing.T) {
+	for _, typ := range testTypes {
+		t.Run(typ, func(t *testing.T) {
+			config := makeMockConfigWatcher()
+			config.deltaResponses = makeDeltaResponses()
+			s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+
+			resp := makeMockDeltaStream(t)
+			// This is a wildcard request since we don't specify a list of resource subscriptions
+			res, err := config.deltaResponses[typ][0].GetDeltaDiscoveryResponse()
+			if err != nil {
+				t.Error(err)
+			}
+			// We only subscribe to one resource to see if we get the appropriate number of responses back
+			resp.recv <- &discovery.DeltaDiscoveryRequest{Node: node, TypeUrl: typ, ResourceNamesSubscribe: []string{res.Resources[0].Name}}
+
+			go func() {
```

Comment:
This pattern seems to be reused between this and the previous test, maybe extract to function?

Reply:
Will do!